Add OpenSUSE to RPM spec file

breed808 opened the pull request:
Referencing issue #1306.
Also some removal of redundant commands/settings. I've built and tested the spec file on Fedora 24/25, CentOS 7 and OpenSUSE Leap 42.1

@ktsaou, you can use this spec file in place of the one currently in use on the VPS build server.

ktsaou commented:
@breed808  thanks!
@alonbl do you have any comments on this?

alonbl commented on netdata.spec.in:
```diff
@@ -10,7 +10,7 @@
 %bcond_with     nfacct   # build with nfacct plugin
 
-%if 0%{?fedora} || 0%{?rhel} >= 7
+%if 0%{?fedora} || 0%{?rhel} >= 7 || 0%{?suse_version} >= 1140
```
this was invalid at first... we should avoid doing undefine and move the upper definition into proper conditionals.

alonbl commented on netdata.spec.in:
```diff
@@ -23,6 +23,11 @@ License:	GPL v3+
 Source0:	http://firehol.org/download/netdata/releases/v@PACKAGE_VERSION@/%{name}-@PACKAGE_VERSION@.tar.xz
 URL:		http://my-netdata.io/
+%if 0%{?suse_version}
+BuildRequires:	automake
```
why do we require automake? we should avoid autoreconf during build

breed808 replied:
You're correct, the automake dependency was not required. I've removed that in a subsequent commit.

alonbl commented on netdata.spec.in:
```diff
@@ -23,6 +23,11 @@ License:	GPL v3+
 URL:		http://my-netdata.io/
+%if 0%{?suse_version}
+BuildRequires:	automake
+BuildRequires:	systemd-rpm-macros
```
is this actually required for build? just curios for what.

anyway, we should put some macro in the prologue to remove this conditional.

breed808 replied:
Correct, `systemd-rpm-macros` is not needed at runtime. Fixed this in a later commit.

alonbl commented on netdata.spec.in:
```diff
@@ -23,6 +23,11 @@ License:	GPL v3+
+%if 0%{?suse_version}
+BuildRequires:	automake
+BuildRequires:	systemd-rpm-macros
+%{?systemd_requires}
```
where do we define this?

breed808 replied:
`%{?systemd_requires} is a macro defined in the systemd-rpm-macros package. See https://build.opensuse.org/package/view_file/openSUSE:Factory/systemd-rpm-macros/macros.systemd?expand=1

alonbl commented on netdata.spec.in:
```diff
@@ -42,15 +47,19 @@ Requires(pre): /usr/sbin/groupadd
+Requires(preun):  systemd-rpm-macros
+Requires(postun): systemd-rpm-macros
+Requires(post):   systemd-rpm-macros
+%else
```
instead of this conditional, please define a macro at prologue with the package name to depend, the less conditionals we have in spec body the easier to maintain.

alonbl commented on netdata.spec.in:
```diff
@@ -62,7 +71,7 @@ so that you can get insights of what is happening now and what just
 
 %prep
-%setup -q -n %{name}-@PACKAGE_VERSION@
+%setup -q
```
how come it works without -n? have you checked rhel/fedora?

breed808 replied:
Yes, have tested with Fedora and CentOS. Read more on the `%setup` macro [here](http://rpm.org/max-rpm-snapshot/s1-rpm-inside-macros.html) to see the defaults.

alonbl replied:
ok

alonbl commented on netdata.spec.in:
```diff
@@ -76,20 +85,18 @@ happened, on your systems and applications.
 %if %{with systemd}
-install -d $RPM_BUILD_ROOT%{_unitdir}
-install -m 644 -p system/netdata.service $RPM_BUILD_ROOT%{_unitdir}/netdata.service
+install -Dm 644 -p system/netdata.service $RPM_BUILD_ROOT%{_unitdir}/netdata.service
```
the install -d is good enough, why have you removed it?

breed808 replied:
I had actually intended the replace the `mkdir` command above it, though wouldn't merging the `install -d` command reduce complexity of the spec file? Happy to change this back if needed.

alonbl replied:
I usually prefer the install -d -m 0755 xxx as explicit, but it is up to you guys... anyway, mkdir should go away.

alonbl replied:
ok, found why install -d -m is to be used... try...
```
cd /tmp
umask 0027
touch a
install -Dm 0664 a /tmp/x1/x2/a
ls -la /tmp/x1
```
notice that x1 and x2 are not what we expect and affected by the umask.

alonbl commented on netdata.spec.in:
```diff
@@ -76,20 +85,18 @@ happened, on your systems and applications.
 # install SYSV init stuff
-mkdir -p $RPM_BUILD_ROOT/etc/rc.d/init.d
-install -m755 system/netdata-init-d \
+install -Dm755 system/netdata-init-d \
```
here there was a mistake, instead of mkdir it should have been install -d

alonbl commented on netdata.spec.in:
```diff
@@ -150,17 +169,16 @@ rm -rf $RPM_BUILD_ROOT
-# node.js config files are JSON (which do not support comments)
-# so, we only supply markdown files
-%{_sysconfdir}/%{name}/node.d/*.md
-
```
not sure why this is removed, trust you guys.

breed808 replied:
Package was failing build on all platforms (Fedora/CentOS/OpenSUSE), as no `/etc/netdata/node.d` directory was ever found in the buildroot.

@ktsaou can you confirm this?

ktsaou replied:
hm... I think we have to create that directory somehow...

ktsaou replied:
Is this something we have to add to the make files?

breed808 replied:
It could be created in the makefile, or during the packaging process with `install -d`. Really up to you to decide.

breed808 replied:
@ktsaou I've run another build, and the `node.d` directory is being created, but not installed.

Would you like me to create another pull request for this, along with the modifications mentioned in #1302 ?

ktsaou replied:
yes please...

alonbl commented on netdata.spec.in:
```diff
@@ -173,6 +191,7 @@ rm -rf $RPM_BUILD_ROOT
 %if %{with systemd}
 %{_unitdir}/netdata.service
 %else
+%dir %{_sysconfdir}/rc.d/init.d
```
package should not own the init.d directory

breed808 replied:
Hmm, I will search for additional info on OpenSUSE's use of the `%dir` macro and get back to you on it.

alonbl commented on netdata.spec.in:
```diff
@@ -101,13 +108,25 @@ install -m755 system/netdata-init-d \
+%service_del_preun netdata.service
+%else
 %systemd_postun_with_restart netdata.service
+%endif
```
if we can also define some macro at prologue to remove these conditionals then we will have simpler sequence.

breed808 commented:
@alonbl thanks for looking through the changes. You mentioned using a macro rather than the multiple conditional checks in a number of comments. Can you give an example of what the macro might resemble? I've gotten by with using conditionals for some time now, so I'm not too familiar with using macros.

Thanks again.

alonbl commented:
BTW: you should also amend your patches, not submit new revisions with fixes. merge should be performed to perfect patches, not sure how to cope with the base, changes, revert of changes etc... this should not go into history.

breed808 commented:
@alonbl for amending the patches, do you mean to hard reset the branch and re-commit with the fixes included?
I was under the impression that once the commits have been pushed, they're not able to be amended or changed in any way, as each commit hash is a result of the previous commit.

alonbl commented:
On 13 December 2016 at 01:01, Ben R <notifications@github.com> wrote:
>
> @alonbl for amending the patches, do you mean to hard reset the branch and re-commit with the fixes included?
> I was under the impression that once the commits have been pushed, they're not able to be amended or changed in any way, as each commit hash is a result of the previous commit.

Git is much more complex/stronger than that.
The workflow is same as sending patches in email, you send patch then
receives comments, then fix your patch and send it again, you do not
send incremental fixes for review.
Git fully automate this process where each patch is a commit, so a
commit must be able to be changed.

In github sequence you should clone a repo, create your own branch,
then write your "patches" as commits, submit to review (pull request),
then when you get comments you should rebase/squash/fixup the existing
commits until no more comments.

In your case you should have at least these two commits:
1. cleanups that are not related to suse.
2. suse specific.

When you (re)push your private branch to your private repo use
--force, notice that github will update this properly in the pull
request.

breed808 commented:
@alonbl, I believe I've rebased them correctly. Many thanks for pointing that out, this is some fairly powerful functionality I haven't been aware of.

As for the conditionals, I have replaced the %post/preun/postun scriptlets with a macro similar to what you suggested.
Should the other conditionals be replaced also? There are still two remaining within the `BuildRequires` and `Requires` sections.

E.G.
```
%if 0%{?suse_version}
...
%define distro_build_requires systemd-rpm-macros
%define distro_requires %{?systemd_requires}
...
%else
...
%endif
...
```

alonbl commented on netdata.spec.in:
```diff
@@ -23,6 +32,9 @@ License:	GPL v3+
 Source0:	http://firehol.org/download/netdata/releases/v@PACKAGE_VERSION@/%{name}-@PACKAGE_VERSION@.tar.xz
 URL:		http://my-netdata.io/
+%if 0%{?suse_version}
+BuildRequires:	systemd-rpm-macros
```
have you reached to a conclusion this is required? if so, please use the same macro conditionals above, and declare macro as empty for other distros.

breed808 replied:
Yes, it is required. I've placed it in the macro conditionals at the top of the page.

alonbl commented on netdata.spec.in:
```diff
@@ -42,9 +54,13 @@ Requires(pre): /usr/sbin/groupadd
 %if %{with systemd}
+%if 0%{?suse_version}
+%{?systemd_requires}
+%else
```
not critical, these can also happen in single conditional at prologue as far as I understand, have you tried it?
```
%xxx() %{expand:\
xxx
yyy
}
```
see examples at /etc/rpm/macros.perl
http://backreference.org/2011/09/17/some-tips-on-rpm-conditional-macros/

breed808 replied:
No, I've had a look through `macros.perl` but I'm still unsure how to apply the conditional.

alonbl commented on netdata.spec.in:
```diff
@@ -146,6 +160,9 @@ rm -rf $RPM_BUILD_ROOT
 
 %dir %{_sysconfdir}/%{name}
+%dir %{_sysconfdir}/%{name}/health.d
+%dir %{_sysconfdir}/%{name}/python.d
```
these belongs to the generic patch :)

breed808 replied:
Moved to the cleanup patch.

ktsaou commented:
Is this ready to be merged?

ktsaou commented:
@alonbl thanks!
@breed808 are you ok to merge it?

breed808 commented:
@ktsaou, yes, no problems here. Good to merge.
@alonbl Thank you for your patience for this pull request, I really appreciate it.

ktsaou commented:
merged!
@facetoe note this.
